✨ Bmp180 driver

50xp50 opened the pull request.

50xp50 committed:
:sparkles: Added basic bmp180 demo

50xp50 committed:
:construction: Rough draft of driver

50xp50 committed:
:sparkles: First revision of bmp180 driver

50xp50 commented:
@kammce Ready for draft review

kammce commented:
Can you change this from a draft to a PR and make sure that CI is passing. Then I'll review.

50xp50 committed:
:memo: Added documentation

50xp50 committed:
:memo: Added datasheet

kammce commented on include/libhal-sensor/multi/bmp180.hpp:
```diff
@@ -0,0 +1,123 @@
+  /**
+   * @brief Stores the temperature and pressure data that are both found when
+   * doing the pressure()
+   */
```
```suggestion
  /**
   * @brief Stores the temperature and pressure data that are both found when
   * calling pressure().
   */
```

kammce commented on include/libhal-sensor/multi/bmp180.hpp:
```diff
@@ -0,0 +1,123 @@
+/**
+ * @brief Driver for the BMP180 barometric pressure/temperature sensor
+ */
+
```
```suggestion
```

kammce commented on include/libhal-sensor/multi/bmp180.hpp:
```diff
@@ -0,0 +1,123 @@
+  struct pressure_results
+  {
+    hal::celsius temperature;
+    float pressure;
```
```suggestion
    // Pressure is in units of "____"
    float pressure;
```

Please add the units for pressure for future usage.

kammce commented on include/libhal-sensor/multi/bmp180.hpp:
```diff
@@ -0,0 +1,123 @@
+   * communication. The calibration data that the sensor uses need to all have
+   * non-boundary values, if any are 0x0000 or 0xFFFF then this is thrown.
+   */
+  explicit bmp180(hal::i2c& p_i2c,
+                  oversampling_rate p_oversampling_setting =
+                    oversampling_rate::ulta_low_power_mode_4500us);
+
```
```suggestion
  bmp180(hal::i2c& p_i2c,
                  oversampling_rate p_oversampling_setting =
                    oversampling_rate::ulta_low_power_mode_4500us);

```

Explicit is only needed when there is a single parameter for a class. C++ allows the assignment operator to act as a constructor if there is just one element which results in a sort of implicit conversion.

50xp50 replied:
Since the second parameter p_oversampling_setting has a default value, then the bmp180 constructor can be called (and treated) as just having the single parameter p_i2c. In this scenario, testing revealed that the assignment operator can be used to declare a bmp180 object. As a result, it's necessary to retain the explicit specifier.

kammce commented on include/libhal-sensor/multi/bmp180.hpp:
```diff
@@ -0,0 +1,123 @@
+   *
+   * @return The temperature in hal::celsius.
+   */
+  [[nodiscard]] hal::celsius temperature();
```
```suggestion
  /**
   * @brief Reads the temperature.
   *
   * @returns the temperature in celsius
   */
  [[nodiscard]] hal::celsius temperature();
```

kammce commented on include/libhal-sensor/multi/bmp180.hpp:
```diff
@@ -0,0 +1,123 @@
+    std::uint16_t ac4{}, ac5{}, ac6{};
+    std::int16_t b1{}, b2{};
+    std::int16_t mb{}, mc{}, md{};
+  };
```
```suggestion
  {
    // The names of the calibration values are taken directly from the datasheet. 
    // The datasheet does not explain what these names mean, so we keep the names as they are defined.
    std::int16_t ac1{}, ac2{}, ac3{};
    std::uint16_t ac4{}, ac5{}, ac6{};
    std::int16_t b1{}, b2{};
    std::int16_t mb{}, mc{}, md{};
  };
```

kammce commented on src/multi/bmp180.cpp:
```diff
@@ -0,0 +1,305 @@
+  if (check_bit == 0)
+    return 1;
+  else
+    return 0;
```
Use curly braces always.

kammce replied:
```suggestion
  constexpr auto check_bit = hal::bit_mask::from(5);
  return hal::bit_extract<check_bit>(buffer[0]);
```

Either one works, but I find the one liner to be easier to work with. May also

kammce commented on src/multi/bmp180.cpp:
```diff
@@ -0,0 +1,305 @@
+};
+
+// helper function to check if sensor has finished its data conversion
+int conversion_status(hal::i2c* p_i2c, hal::byte p_address)
```
```suggestion
bool conversion_status(hal::i2c* p_i2c, hal::byte p_address)
```
Consider returning a bool since this function only returns 1 or 0.

kammce commented on src/multi/bmp180.cpp:
```diff
@@ -0,0 +1,305 @@
+  int conversion_finished = 0;
+  while (conversion_finished == 0) {
+    conversion_finished = conversion_status(p_i2c, p_address);
+  }
```
```suggestion
  bool conversion_finished = 0;
  while (conversion_finished == 0) {
    conversion_finished = conversion_status(p_i2c, p_address);
  }
```

kammce commented on src/multi/bmp180.cpp:
```diff
@@ -0,0 +1,305 @@
+  variables.x2 = (p_mc << 11) / (variables.x1 + p_md);
+  variables.b5 = variables.x1 + variables.x2;
+  variables.temperature = (variables.b5 + 8) >> 4;
+
```
Add a comment that also describes why you aren't using `bit_modify`, because this is written directly in the datasheet, and where in the datasheet this comes from. This way its obvious for someone to look at this code and check your work.

kammce commented on src/multi/bmp180.cpp:
```diff
@@ -0,0 +1,305 @@
+                            uncompensated_output_data_register(),
+                            hal::never_timeout());
+  std::int32_t uncompensated_temperature =
+    (uncompensated_data_buffer[0] << 8) | uncompensated_data_buffer[1];
```
Is this supposed to be a signed 32-bit integer? If so, are the 16-bits you are ORing together also signed? If so, should we be doing some sort of sign extension on this?

50xp50 replied:
This is as its referenced and described in the datasheet, with no further explanations on their behalf. Keeping it as is.

kammce commented on src/multi/bmp180.cpp:
```diff
@@ -0,0 +1,305 @@
+  variables.temperature = (variables.b5 + 8) >> 4;
+
+  hal::celsius temperature_in_celsius =
+    static_cast<float>(variables.temperature) / 10.0;
```
```suggestion
  hal::celsius temperature_in_celsius =
    static_cast<float>(variables.temperature) / 10.0f;
```

Use suffix f to make sure a number is float not double.

kammce commented on src/multi/bmp180.cpp:
```diff
@@ -0,0 +1,305 @@
+      break;
+    default:
+      m_maximum_samples = 38;
+      break;
```
```suggestion
    case oversampling_rate::ultra_high_resolution_mode_25500us:
      [[fallthrough]];
    default:
      m_maximum_samples = 38;
      break;
```

kammce replied:
Since they are the same, might as well fallthrough.

kammce commented on src/multi/bmp180.cpp:
```diff
@@ -0,0 +1,305 @@
+
+  // tell sensor to start conversion for pressure
+  hal::byte control_setting =
+    (static_cast<hal::byte>(m_oversampling_setting) << 6) | 0x34;
```
```suggestion
  hal::byte control_setting =
    (hal::value(m_oversampling_setting) << 6) | 0x34;
```

Another option is hal::value from the `#include <libhal-util/enum.hpp>` header.

kammce commented on include/libhal-sensor/multi/bmp180.hpp:
```diff
@@ -0,0 +1,123 @@
+   * setting.
+   * @return pressure_results which contains the pressure and temperature
+   */
+  [[nodiscard]] pressure_results pressure(int sample_amount = 1);
```
Can you explain how its different and how the sample_amount works? For example, there is not explanation in that the sample amount is clamped to be between 1 and the max amount. There is also no documentation on what the sample amount maxes are unless you read the implementation file. Can you add that somewhere in this description or where ever else it may be relevant.

kammce commented on src/multi/bmp180.cpp:
```diff
@@ -0,0 +1,305 @@
+  };
+
+  hal::byte oversampling_setting =
+    static_cast<hal::byte>(m_oversampling_setting);
```
Since you use this a few times, might as well create the variable above line 226. The compiler will optimize this for you, but its easier to read if we know we are reusing the same variable multiple times above.

kammce commented on src/multi/bmp180.cpp:
```diff
@@ -0,0 +1,305 @@
+    hal::write(*m_i2c, m_address, sensor_config_buffer, hal::never_timeout());
+    while (conversion_finished == 0) {
+      conversion_finished = conversion_status(m_i2c, m_address);
+    }
```
```suggestion
  for (int i = 0; i < sample_amount; i++) {
    hal::write(*m_i2c, m_address, sensor_config_buffer, hal::never_timeout());
    int conversion_finished = 0;
    while (conversion_finished == 0) {
      conversion_finished = conversion_status(m_i2c, m_address);
    }
```

Keep variables close the scope where they are used. And refrain from uninitialized variables.

kammce replied:
It may even be better to simply put this into its own function since you tend to need it a few times in the code. The function will spin until the conversion is done.

kammce commented on src/multi/bmp180.cpp:
```diff
@@ -0,0 +1,305 @@
+    std::int32_t uncompensated_pressure =
+      ((uncompensated_data_buffer[0] << 16) |
+       (uncompensated_data_buffer[1] << 8) | uncompensated_data_buffer[2]) >>
+      (8 - oversampling_setting);
```
Can you add a comment that references where in the datasheet this equation comes from? There is no way to know this is correct, otherwise.

kammce commented on src/multi/bmp180.cpp:
```diff
@@ -0,0 +1,305 @@
+  average_uncompensated_pressure =
+    summed_uncompensated_pressures / sample_amount;
+
+  // convert the raw pressure data into Pascals in steps of 1Pa
```
Add a comment explain why you did not use bit_modify here and where in the datasheet these equations are located.

kammce replied:
And that this is written exactly as it was spelled out in the datasheet.

kammce commented on src/multi/bmp180.cpp:
```diff
@@ -0,0 +1,305 @@
+{
+  hal::byte reset_setting = 0xB6;
+  std::array<hal::byte, 2> sensor_config_buffer = { soft_reset_register()[0],
+                                                    reset_setting };
```
```suggestion
  hal::byte reset_setting = 0xB6;
  std::array<hal::byte, 2> sensor_config_buffer = { soft_reset_register()[0],
                                                    reset_setting };
```

Since the reset_setting value is actually a constant, consider having soft_reset_register() return back an array of 2 bytes with the reset setting in the right place. Then this call will just be a single `hal::write` with soft_reset_register in place of sensor_config_buffer.

kammce reviewed (changes requested):
Great job Xander! This looks great!

50xp50 committed:
:art: Improved documentation and functions

50xp50 committed:
:art: Updated code to reflect PR suggestions

- Brokeout the temperature and pressure calculations into functions.
- Added unit tests for the calculation functions using the datasheet sample data.
- Updated helper function to wait for conversion.
- Updated/fixed comments and added more clarifying information.
- Some additional tidying-up and minor edits

kammce commented on include/libhal-sensor/multi/bmp180.hpp:
```diff
@@ -24,13 +24,12 @@ namespace hal::sensor {
-   * @brief Oversampling rate defined with the mode names as well as sample
-   * fetch times.
+   * @brief Stores the oversampling rate which is defined with the mode names as
+   * well as sample fetch times.
```
```suggestion
   * @brief Defines the oversampling rate with names containing the mode as
   * well as sample fetch times.
```

I wouldn't consider an enum class as "storing" stuff. I think this is what you meant in your comment.

kammce commented on include/libhal-sensor/multi/bmp180.hpp:
```diff
@@ -46,46 +45,62 @@ class bmp180
+   * ulta_low_power_mode_4500us = 3 internal samples by sensor
+   * standard_mode_7500us = 5 internal samples by sensor
+   * high_resolution_mode_13500us = 9 internal samples by sensor
+   * ultra_high_resolution_mode_25500us = 17 internal samples by sensor
```
```suggestion
   * RMS noise from the sensor. See documentation for `oversampling_rate` enumeration class to know the internal samples taken for each mode.
   * 
```

kammce replied:
No need to spell the same thing again if we can refer to the original source.

kammce commented on include/libhal-sensor/multi/bmp180.hpp:
```diff
@@ -46,46 +45,62 @@ class bmp180
+   * ulta_low_power_mode_4500us = 214 samples
+   * standard_mode_7500us = 128 samples
+   * high_resolution_mode_13500us = 72 samples
+   * ultra_high_resolution_mode_25500us = 38 samples
```
```suggestion
   * 
   * The maximum samples per oversampling rate setting is as follows:
   * 
   *     - ulta_low_power_mode_4500us = 214 samples
   *     - standard_mode_7500us = 128 samples
   *     - high_resolution_mode_13500us = 72 samples
   *     - ultra_high_resolution_mode_25500us = 38 samples
```

kammce commented on src/multi/bmp180.cpp:
```diff
@@ -1,16 +1,20 @@
-// their payloads
+// Helper functions for creating arrays containing the different register
+// addresses in their payloads.
 auto device_id_register()
```
```suggestion
// Helper functions for creating arrays containing the different register
// addresses in their payloads.
constexpr auto device_id_register()
```

Make all of the helper functions that just return constant information constexpr to give the compiler more of an opportunity to optimize these.

kammce commented on src/multi/bmp180.cpp:
```diff
@@ -31,84 +35,76 @@ auto uncompensated_output_data_register()
 {
-  return std::to_array<hal::byte>({ 0xE0 });
+  hal::byte reset_address = 0xE0;
+  hal::byte reset_setting = 0xB6;
```
```suggestion
  constexpr hal::byte reset_address = 0xE0;
  constexpr hal::byte reset_setting = 0xB6;
```
Make all constants constexpr when you can.

kammce commented on src/multi/bmp180.cpp:
```diff
@@ -31,84 +35,76 @@ auto uncompensated_output_data_register()
+    constexpr auto conversion_status_bit = hal::bit_mask::from(5);
+    conversion_finished = !(hal::bit_extract<conversion_status_bit>(buffer[0]));
+  }
+}
```
Maybe a better name would be `conversion_busy`?

```suggestion
  bool conversion_busy = true;
  while (conversion_busy) {
    auto status = hal::write_then_read<1>(*p_i2c,
                         p_address,
                         control_measurement_register(),
                         buffer,
                         hal::never_timeout())[0];
    constexpr auto conversion_status_bit = hal::bit_mask::from(5);
    conversion_busy = hal::bit_extract<conversion_status_bit>(status);
  }
}
```

Use `true` and `false` for bools. Using integers for booleans can be confusing.
Also, `hal::write_then_read<1>` has a means to return an array, so you can remove the std::array buffer.

kammce commented on src/multi/bmp180.cpp:
```diff
@@ -196,97 +191,77 @@ bmp180::bmp180(hal::i2c& p_i2c, oversampling_rate p_oversampling_setting)
-    (static_cast<hal::byte>(m_oversampling_setting) << 6) | 0x34;
+  // Prepare buffer to configure the sensor for pressure data gathering.
+  std::int16_t oversampling_setting = hal::value(m_oversampling_setting);
+  hal::byte control_setting = (oversampling_setting << 6) | 0x34;
```
```suggestion
  hal::byte control_setting = (oversampling_setting << 6) | 0x34;
```

Why do we OR with 0x34? If this is something in the datasheet, its important to call it out so its obvious to the reader.

kammce commented on src/multi/bmp180_internal.hpp:
```diff
@@ -0,0 +1,72 @@
+    std::int32_t b3, b5, b6;
+    std::uint32_t b4, b7;
+    std::int32_t temperature, pressure;
+    };
```
Can you run clang-format on this file?

kammce commented on src/multi/bmp180_internal.hpp:
```diff
@@ -0,0 +1,72 @@
+        return variables;
+    }
+
+}
```
Add a newline to the end of this file.

kammce commented on tests/multi/bmp180.test.cpp:
```diff
@@ -0,0 +1,106 @@
+
+#include <libhal-sensor/multi/bmp180.hpp>
+
+#include "../../src/multi/bmp180_internal.hpp"
```
```suggestion
#include "multi/bmp180_internal.hpp"
```

You should only need to do this for the include.

kammce reviewed (changes requested):
Great work on the tests! Just a few last things, and we can merge this in.

50xp50 committed:
:art: Update to reflect PR requested changes

kammce commented on src/multi/bmp180_internal.hpp:
```diff
@@ -6,67 +6,83 @@
+// These calculations are done the way they are without bit_modify as its
+// following the exact algorithm for conversion given in figure 4 of the
+// datasheet.
+inline computation_variables calculate_true_temperature(
```
You could actually set this to be contexpr since there is nothing inside that cannot be computed at compile time.

kammce replied:
I'll just resolve this now actually.

kammce reviewed (approved):
Looks great @50xp50 ! This should be good to submit 😄 Cheers!

50xp50 committed:
:art: Tweak to bmp_internal functions
